src/rnnlm/sample_a_word.cc

```diff
@@ -0,0 +1,432 @@
+    ed += pow(pdf_est[word] - pdf[word], 2); 
+  }
+  ed = pow(ed, 0.5);
+  std::cout << "Run " << iters << " times, e distance (expect < 0.05) is " << ed << std::endl;
+  std::cout << "Number of words OOV : " << count_nons << std::endl;
+}
+
+// Test the read-in language model
+void NgramModel::TestReadingModel() {
+  std::cout << "Testing model reading part..."<< std::endl;
+  std::cout << "Vocab size is: " << vocab_size_ << std::endl;
+  std::cout << "Ngram_order is: " << ngram_order_ << std::endl;
+  assert(probs_.size() == counts_.size());
+  for (int32 i = 0; i < ngram_order_; i++) {
+    int32 size_ngrams = 0;
+    std::cout << "Test: for order " << (i + 1) << std::endl;
+    std::cout << "Expected number of " << (i + 1) << "-grams: " << counts_[i] << std::endl;
+    for (auto it1 = probs_[i].begin(); it1 != probs_[i].end(); ++it1) {
+      HistType h(it1->first);
+      for (auto it2 = (probs_[i])[h].begin(); it2 != (probs_[i])[h].end(); ++it2) {
+        size_ngrams++; // number of words given
+      }
+    }
+    std::cout << "Read in number of " << (i + 1) << "-grams: " << size_ngrams << std::endl;
+  }
+  std::cout << "Assert sum of unigram probs equal to 1..." << std::endl;
+  float prob_sum = 0.0;
+  int32 count = 0;
+  for (auto it1 = (probs_[0]).begin(); it1 != (probs_[0]).end();++it1) {
+    HistType h(it1->first);
+    for (auto it2 = (probs_[0])[h].begin(); it2 != (probs_[0])[h].end(); ++it2) {
+      prob_sum += 1.0 * pow(10.0, (it2->second).first);
+      count++;
+    }
+  }
+  std::cout << "Number of total words: " << count << std::endl;
+  std::cout << "Sum of unigram probs equal to " << prob_sum << std::endl;
+
+  std::cout << "Assert sum of bigram probs given a history equal to 1..." << std::endl;
+  prob_sum = 0.0;
+  auto it1 = probs_[1].begin();
+  HistType h(it1->first);
+  for (auto it = vocab_.begin(); it != vocab_.end(); ++it) {
+    auto it2 = probs_[1][h].find(it->second);
+    if (it2 != probs_[1][h].end()) {
+      prob_sum += 1.0 * pow(10, (it2->second).first);
+    } else {
+      prob_sum += pow(10, GetProb(2, it->second, h));
+    }
+  }
+  std::cout << "Sum of bigram probs given a history equal to " << prob_sum << std::endl;
+
+}
+
+// Read histories of integers from a file
+void NgramModel::ReadHistories(char* file) {
+ std::ifstream data_input(file);
+  if (!data_input.is_open()) {
+    std::cerr << "error opening '" << file
+              << "' for reading\n";
+    exit(1);
+  }
+  std::string line;
+  std::cout << "Start reading histories..." << std::endl;
+  while (getline(data_input, line)) {
+    std::istringstream is(line);
+    std::istream_iterator<std::string> begin(is), end;
+    std::vector<std::string> tokens(begin, end);
+    HistType history;
+    int32 word;
+    for (int32 i = 0; i < tokens.size(); i++) {
+      auto it = vocab_.find(tokens[i]);
+      if (it != vocab_.end()) {
+        word = it->second;
+      } else {
+        std::string word_s = "<unk>";
```

Comment:
add a class member called unk_, which by default is "<unk>", but could be set
in the constructor.